File: Leetcode_133.java
class Solution {
    private HashMap<Node, Node> visited = new HashMap<>();

    public Node cloneGraph(Node node) {
        if (node == null) return null;                         // base case

        if (visited.containsKey(node))                         // already cloned?
            return visited.get(node);

        Node clone = new Node(node.val, new ArrayList<>());    // clone node
        visited.put(node, clone);                              // mark as visited

        for (Node nei : node.neighbors)                        // clone neighbors
            clone.neighbors.add(cloneGraph(nei));

        return clone;                                           // return clone root
    }
}

File: Leetcode_133.cpp
class Solution {
public:
    unordered_map<Node*, Node*> visited;

    Node* cloneGraph(Node* node) {
        if (!node) return nullptr;                               // base case

        if (visited.count(node))                                 // already cloned?
            return visited[node];

        Node* clone = new Node(node->val);                       // clone node
        visited[node] = clone;                                   // mark as visited
        
        for (Node* nei : node->neighbors)                        // clone neighbors
            clone->neighbors.push_back(cloneGraph(nei));

        return clone;                                            // return clone root
    }
};
